README
NOTE: This repo is solely an archive of the final version 1.4.9 by anonNL made in 2011, in the aim of defining the least common denominator for a working implementation. https://web.archive.org/web/20111207074836/http://digital-haze.net/files/vnds/vnds/vnds-1.4.9.tar.gz

A maintained version for Nintendo DS can be found below:

https://github.com/ds-homebrew/vnds

Note that anonNL indicates that jake's contributions starting with 1.5.0 (lost) are a separate branch based on 1.4.7 (lost)
so these commits are actually older and precede the true latest version 1.4.9 which merges changes from jake's 1.5.3 and 
provides newer patches, confusingly. (as shown by the crossed out links and the 1.4.9 ChangeLog) https://web.archive.org/web/20140218222855/digital-haze.net/projects/vnds.html

---

Additional documentation is available in the manual folder

Installation:
    Extract the .tar.gz onto the root of the SD card so the path is 
    /vnds/. To install a visual novel, extract the tar.gz in /vnds/novels, so
    it has its own folder (ie. /vnds/novels/tsukihime). Sound files should
    end up in /vnds/novels/<game folder>/sound/.

NOTICE:
    * any save prior to version 1.3.1 needs to be ran though the sav2xml.py
        utility in the tools/ folder. requires python(http://python.org) to run
    * alternatively: http://digital-haze.net/files/vnds/fixsave.php

Running:
    Should be easy enough

Troubleshooting:
    * Make sure the path is correct
    * If you run it and you get stuck at a white screen, launch it with an
        alternative launcher, such as dsorganize or dschannels.

Contact:
    forum: http://digital-haze.net/ch/vnds
    IRC:   irc.rizon.net #vnds

Credits:
    Programming: Jake Probst, anoNL



ChangeLog
* 1.4.9 13/01/10
	anoNL:
		* png error handling (ignore spurious warnings/crash on invalid png)
		* "text ~" inserts blank lines again

* 1.4.8 21/10/09
    anoNL:
    	* backported new functions up to 1.5.3 (excluding wifi + .novel fmt)
        * added support for unpacking .novel files to NovelManager
        * ansi color code support in the text renderer
        * increased sprite limit from 4 to 9
        * waking up from sleep mode was glitchy

* 1.5.3 22/06/09
    Jake:
        * fixed bug with undefined variables in conditionals
        * string variables need to be ""`d (setvar/if)
        * wifi doesnt hang forever if it can't connect

* 1.5.2 03/06/09
    Jake:
        * fixed problems with music with too high a bitrate

* 1.5.1 16/05/09
    Jake:
        * .novel thumbnail.png select screen crash bug fixed

* 1.5.0 04/05/09
    Jake:
        * .novel support
        * download novels directly onto ds
        * redid FIFO stuff
        * 'if var == 0' works with unset variables again
    Dilt:
        * cleartext command

* 1.4.7 23/04/09
    anoNL:
        * a glitch in the image cache caused corrupted images to display
        * scripts starting with a choice command didn't load properly
        * an incorrect linenumber would be saved after following a goto
        * text commands starting with a space were rendered incorrectly
        * large (>1KB) save files sometimes caused stack overflows

* 1.4.6 10/04/09
    Jake:
        jump takes label argument

* 1.4.5 09/04/09
    anoNL:
        * bg changes sometimes didn't display
        * clock didn't update
        * assigning values from local vars to globals didn't work
        * bgload now has a second param to control the fade speed
        * bgm volume changes take effect immediately
        * comparing ints with strings works as expected now
        * support for {$var} notation
        * better autofade algorithm
        * updated packaged tools

* 1.4.4 24/03/09
    Jake:
        * variables can be inserted into most commands
    
    anoNL:
        * fixed sound skipping
        * string variables fixed

* 1.4.3 21/03/09
    Jake:
        * variables can be assigned to other variables
        * label/goto commands added
        * updated to libnds 1.3.2

* 1.4.2 25/01/09
    anoNL:
    	* global vars were reset when loading a save file.
    	* changing the skin via config.ini didn't work.
    	* date wasn't saved properly in save files. 
    	* missing resources no longer generate warning messages.
    	* fast reading by holding Y is now a lot faster.
    	* background fades more intelligent (fade to black on bg change)

* 1.4.1 10/01/08
    anoNL:
        * Fixed some graphical glitches in the GUI.
        * Backlight control could only turn the screen off on the original DS.
        * PNG loading crashed on corrupt or wide (> 256 pixels) images.
        * Sound didn't unmute when returning to the main menu.
        * Tapping the screen advanced the script even in choice-mode.
        * Upgraded to libnds 1.3.1 internally

* 1.4.0 24/12/08
    anoNL:
        * Fixed a memory leak and a wild pointer in png loader.
        * Prints an error instead of crashing if an image is corrupted.
        * Large parts of the program have been rewritten, it's now much more stable.
        * Decreased png loading time by 10-25%
        * ZIP file entries are cached to decrease novel load times.
        * Scans ahead to preload the next image.
        * Uses a modified version of libfat to decrease seek-times in large files.
        * Crossfades between bgload/setimg commands.
        * Added user-changeable settings: fontsize/textspeed/volume (music & sound)
        * Program is skinnable, change the current skin via config.ini
        * Novels can contain partial skins that override the user's skin.

    Jake:
        * touching screen advances text
        * smoother load/save image animation
        * touchpad scrolling downwards improved

* 1.3.7 29/06/08:
    anoNL:
        * bugfix: variables weren't being initialized in some cases.

* 1.3.6 26/06/08:
    anoNL:
        * bugfix: spacing between items in the novel selection list was
            incorrect
        * bugfix: writing outside of array bounds in the script reader
        * change the font-size for a novel by adding "fontsize=14" (or any
            another integer) to info.txt

* 1.3.5 02/06/08:
    Jake:
        * fixed 3 small bugs in loading saves
        * made `endscript` actually do something (returns to titlescreen)

* 1.3.4 29/05/08:
    Jake:
        * fixed a bug in loading saves/skipping text.
        * something happened to make True Remembrance work (diff doesnt
            explain much).

* 1.3.3 25/05/08:
    Jake:
        * fixed problem with time not changing
        * .sav keeps track of game state to make loading faster
        * does nothing when attempting to load an empty save
        * loads script in 4k blocks
        * left/right moves to first/last option in select screen/choices

* 1.3.2 23/05/08:
    anoNL:
        * bugfix: ArchiveFileHandle.Read() didn't work as expected (didn't
            update the read pointer afterwards)
        * bugfix: removed file io from interrupt handler
        * bugfix: texture streaming bug in the save/load menu
        * rewrote text cache to use freetype's built-in caching system
        * added option to interrupt delay commands by pressing A, B or the
            touchscreen
        * skip scene functionality, hold R and press Y to skip ahead to the
            next scene/choice
        * added support for AAC format (for sound effects only)


* 1.3.1 23/05/08:
    Jake:
        * fixed linebreaks in choices
        * string variables
        * variable placement in commands
        * moved save format over to .xml (converter provided)
        * less efficient, but more stable sleep mode
        * collision detection more efficient
        * new save theme

* 1.3.0 21/05/08:
    anoNL:

src/arm9/vnds.cpp
#include "vnds.h"

#include "graphics_engine.h"
#include "res.h"
#include "script_engine.h"
#include "sound_engine.h"
#include "text_engine.h"
#include "vnchoice.h"
#include "maintextpane.h"
#include "tcommon/dsunzip.h"
#include "tcommon/gui/gui_common.h"

using namespace std;

VNDS::VNDS(NovelInfo* novelInfo) {
	quit = false;
	delay = 0;
	waitForInput = false;

	srand(time(NULL));

    gui = new GUI();
	gui->VideoInit();

	//Show loading image
	u16* temp = new u16[256*384];
	skinLoadImage("%s/loading_bg", temp, 256, 384);
	DC_FlushRange(temp, 256*384*2);
	dmaCopyHalfWordsAsynch(1, temp, gui->sub_bg, 256*192*2);
    dmaCopyHalfWordsAsynch(2, temp + 256*192, gui->main_bg, 256*192*2);
    while (dmaBusy(1) | dmaBusy(2));
    delete[] temp;

    //Init GUI and subsystems
    textEngine = new TextEngine(gui, this, novelInfo->GetFontSize());
    textEngine->SetBackgroundColor(skin->GetBackgroundColor());
    gui->PushScreen(textEngine);

    //Copy texture into VRAM, but don't save a copy in TextEngine
    u16* textureI = new u16[256*256];
    skinLoadImage("%s/ingame_tex", textureI, 256, 256);
    textEngine->SetTexture(textureI);
    textEngine->SetTexture(NULL);
    delete[] textureI;

    //Start
    chdir(novelInfo->GetPath());

    vnLog(EL_verbose, COM_CORE, "VNDS init ok");
}
VNDS::~VNDS() {
	if (gui) delete gui;

	chdir("../..");
}

void VNDS::Reset() {
	textEngine->Reset();
	graphicsEngine->Reset();
	scriptEngine->Reset();
	soundEngine->Reset();

	globals.clear();
	variables.clear();

	loadGlobal(this);
	waitForInput = false;
	delay = 0;
}

void VNDS::Idle() {

	//Preload the next png image
	#ifdef IMG_LOAD_PROFILE
		iprintf("\x1b[s\x1b[0;28H%s\x1b[u",
				(pngStream->IsImageReady() ? "100%" : "    "));
	#endif

	if (pngStream->IsStreaming()) {
		#ifdef IMG_LOAD_PROFILE
			if (pngStream->height > 0) {
				iprintf("\x1b[s\x1b[0;28H%3d%%\x1b[u", 100 * pngStream->line / pngStream->height);
			}
		#endif

		int t = 0;
		while (t < PNG_STREAM_LINES && pngStream->ReadLine()) {
			t++;
		}
	} else if (!pngStream->IsImageReady()) {
		for (int n = 0; n < PNG_STREAM_LINES; n++) {
			Command cmd = scriptEngine->GetCommand(n);
			if (cmd.id == SETIMG) {
			    char path[MAXPATHLEN];
			    sprintf(path, "foreground/%s", cmd.setimg.path);
			    if (!graphicsEngine->IsImageCached(path)) {
					pngStream->Start(foregroundArchive, path);
					break;
			    }
			}
		}
	}
}

bool VNDS::IsWaitingForInput() {
	return delay <= 0 && gui->GetActiveScreen() == textEngine
		&& !textEngine->GetChoiceView()->IsActive() && waitForInput
		&& textEngine->GetTextPane()->IsFullTextDisplayed();
}
int VNDS::GetDelay() {
	return delay;
}

void VNDS::Update() {
	if (gui->GetActiveScreen() != textEngine) {
		return;
	}

	u32 down = keysDown();
	u32 held = keysHeld();

    u32 anyKey = KEY_A|KEY_B|KEY_X|KEY_Y|KEY_START|KEY_SELECT|KEY_L|KEY_R|KEY_TOUCH;

	if (delay > 0) {
		delay--;
        if (down & anyKey) {
            delay = 0;
        }
		Idle();
	} else if (held & KEY_R) {
		//R held

		if (!textEngine->GetChoiceView()->IsActive()) {
			if (down & KEY_Y) {
			    int oldMode = (REG_DISPCNT);
			    const int fadeFrames = 8;

			    u16* loadingI = new u16[256*192];
			    bool imageOK = skinLoadImage("../../%s/skipload_bg", loadingI, 256, 192);
				if (imageOK) {
					memset(loadingI+(176<<8), 0, (16<<8)*2);
				    vramSetBankD(VRAM_D_LCD);
				    dmaCopy(loadingI, VRAM_D, 256*192*sizeof(u16));
					vramSetBankD(VRAM_D_MAIN_BG_0x06020000);

					REG_BG2CNT = BG_BMP16_256x256 | BG_BMP_BASE(8) | BG_PRIORITY(1);
					videoSetMode(DEFAULT_VIDEO_MODE | DISPLAY_BG2_ACTIVE);

					REG_BLDALPHA = 0 | (0x10 << 8);
					REG_BLDCNT = BLEND_ALPHA | BLEND_SRC_BG2 | BLEND_DST_BG0 | BLEND_DST_BG3;
				    for (int n = 0; n <= fadeFrames; n++) {
				        int a = 0x0B * n / fadeFrames;
				        int b = 0x10 - a;
				        REG_BLDALPHA = a | (b << 8);
				        swiWaitForVBlank();
				    }
				}
				delete[] loadingI;

				scriptEngine->QuickRead();

				if (imageOK) {
				    for (int n = 0; n <= fadeFrames; n++) {
				        int a = 0x10 - 0x0B * n / fadeFrames;
				        int b = 0x10 - a;
				        REG_BLDALPHA = a | (b << 8);
				        swiWaitForVBlank();
				    }
				    REG_BLDCNT = BLEND_NONE;

				    videoSetMode(oldMode);
				}
			} else if (down & KEY_A) {
				vnLog(EL_message, COM_CORE, "Heap Size: %d.%03d.%03d", HEAP_SIZE/1000000,
						(HEAP_SIZE/1000)%1000, HEAP_SIZE%1000);
			} else if (down & KEY_B) {
			    SaveScreenShot();
			}
		}

		if (down & KEY_X) {
			textEngine->ShowSaveMenu();
		}
	} else if (held & KEY_L) {
		//L held

		if (down & KEY_X) {
			textEngine->ShowLoadMenu();
		}
	} else {
		//No buttons held
		if (down & (KEY_X|KEY_START)) {
			textEngine->ShowMenu();
		} else if (gui->GetActiveScreen() == textEngine && !textEngine->GetChoiceView()->IsActive()) {
			if (!waitForInput) {
				Continue(held & KEY_Y);
			} else {
				if (!textEngine->GetTextPane()->IsFullTextDisplayed()) {
					if ((down & anyKey) || (held & KEY_Y)) {
						textEngine->GetTextPane()->DisplayFullText();
					}
				} else if ((down & (KEY_A|KEY_B)) || (held & KEY_Y)) {
					Continue(held & KEY_Y);
				}
			}
		} else {
			Idle();
		}
	}
}

void VNDS::Continue(bool quickread) {
	delay = 0;
	waitForInput = false;
	scriptEngine->ExecuteNextCommand(quickread);
}

void VNDS::Quit() {
	quit = true;
}

bool zipError;

void onZipProgress(int val, int max, const char* message) {
	if (max >= 0) {
		if (message) {
			int messageL = strlen(message);
			iprintf("\x1b[16;0H%*s%s%*s", (32-messageL+1)/2, " ", message, (32-messageL)/2, " ");
		} else {
			iprintf("\x1b[16;0H                               ");
		}
		if (max > 0) {
			iprintf("\x1b[17;14H%3d%%", 100*val/max);
		} else {
			iprintf("\x1b[17;0H                               ");
		}
	} else {
		zipError = true;

		consoleDemoInit();
		BG_PALETTE_SUB[0]   = BIT(15)|RGB15(0, 0, 31);
		BG_PALETTE_SUB[255] = BIT(15)|RGB15(31, 31, 31);

	    const char* warning =
	    "--------------------------------"
	    "           Fatal Error          "
	    "--------------------------------"
	    "                                "
		" The zip file contains a        "
		" compressed entry, but only     "
	    " uncompressed zip files are     "
	    " supported.                     "
	    "                                "
	    "--------------------------------";
	    iprintf(warning);

	    waitForAnyKey();
	}
}

void VNDS::Run() {
	quit = false;

	//Open archives
    consoleClear();
    zipError = false;
    foregroundArchive = openArchive("foreground", "foreground/", &onZipProgress);
    if (!zipError) {
    backgroundArchive = openArchive("background", "background/", &onZipProgress);
    if (!zipError) {
    scriptArchive = openArchive("script", "script/", &onZipProgress);
    if (!zipError) {
    soundArchive = openArchive("sound", "sound/", &onZipProgress);
    if (!zipError) {
    consoleClear();

    //Load font
    if (fexists("default.ttf")) {
    	textEngine->GetTextPane()->SetFont("default.ttf");
    	textEngine->GetChoiceView()->SetFont("default.ttf");
    	vnLog(EL_verbose, COM_TEXT, "Custom font (default.ttf) found.");
    }

    //Init state
    loadGlobal(this);

    scriptEngine = new ScriptEngine(this, scriptArchive);
    scriptEngine->SetScriptFile("script/main.scr");

    graphicsEngine = new GraphicsEngine(this, (u16*)BG_BMP_RAM_SUB(2), backgroundArchive, foregroundArchive);
    graphicsEngine->Flush(true);

    soundEngine = new SoundEngine(soundArchive);
	pngStream = new PNGStream(soundEngine);

	textEngine->Activate();

    //Main loop
	while (!quit) {
		gui->Update();
		Update();

		if (!quit) {
			gui->Draw();
			soundEngine->Update();
			swiWaitForVBlank();
		}
	}

	fadeBlack(16);

	//Cleanup
	delete scriptEngine;
	scriptEngine = NULL;

	delete graphicsEngine;
	graphicsEngine = NULL;

	delete soundEngine;
	soundEngine = NULL;

	delete pngStream;
	pngStream = NULL;

	//Close ZIP archives
	closeArchive(soundArchive);
	soundArchive = NULL;
	}
	closeArchive(scriptArchive);
	scriptArchive = NULL;
	}
	closeArchive(backgroundArchive);
	backgroundArchive = NULL;
	}
	closeArchive(foregroundArchive);
	foregroundArchive = NULL;
	}

	resetVideo();
}

void VNDS::SaveScreenShot() {
	vnLog(EL_message, COM_CORE, "Saving screenshot...");

	bool wasMuted = soundEngine->IsMuted();
	soundEngine->SetMuted(true);

	vramSetBankD(VRAM_D_LCD);
	setupCapture(3);
	gui->Draw();
	waitForCapture();

	u16* temp = new u16[256*384];
	memcpy(temp, gui->sub_bg, 256*192*2);
	memcpy(temp+256*192, VRAM_D, 256*192*2);

    char filename[32];
    time_t ti = time(NULL);
    tm* tm = localtime(&ti);
    unsigned int timestamp = tm->tm_sec + 60*(tm->tm_min + 60*(tm->tm_hour + 24*(tm->tm_yday + 366 * tm->tm_year)));
    sprintf(filename, "%u.png", timestamp);

    if (pngSaveImage(filename, temp, 256, 384)) {
		vnLog(EL_message, COM_CORE, "Screenshot saved successfully.");
	} else {
		vnLog(EL_message, COM_CORE, "Error saving screenshot.");
	}
	delete[] temp;

	soundEngine->SetMuted(wasMuted);
}

void VNDS::SetDelay(int d) {
	delay = d;
}
void VNDS::SetWaitForInput(bool b) {
	waitForInput = b;
}

void VNDS::SetVariable(const char* name, char op, const char* value) {
	SetVar(variables, name, op, value);
}
void VNDS::SetGlobal(const char* name, char op, const char* value) {
	SetVar(globals, name, op, value);
	saveGlobal(this);
}
void VNDS::SetVar(map<std::string, Variable>& map,
		const char* name, char op, const char* value)
{
	if (op == '~') {
		map.clear();
		return;
	}

    Variable var(value);
    if (variables.count(value) != 0) {
        var = variables[value];
    } else if (globals.count(value) != 0) {
        var = globals[value];
    }
	Variable target = map[name];

	VarType inferredType = VT_int;
	if (target.type == VT_string || var.type == VT_string) {
		inferredType = VT_string;
	}
	target.type = inferredType;

	switch (inferredType) {
	case VT_int:
		switch (op) {
		case '+':
			target.intval += var.intval;
			break;
		case '-':
			target.intval -= var.intval;
			break;
		case '=':
			target.intval = var.intval;
			break;
		default:
			vnLog(EL_warning, COM_CORE, "setvar :: Unsupported operator '%c' for target type int", op);
			return;
		}

		sprintf(target.strval, "%d", target.intval);
		break;
	case VT_string:
		int offset;
		switch (op) {
		case '+':
			offset = strlen(target.strval);
			strncpy(target.strval+offset, var.strval, VAR_STRING_LENGTH-offset-1);
			target.strval[VAR_STRING_LENGTH-1] = '\0';
			break;
		case '=':
			strncpy(target.strval, var.strval, VAR_STRING_LENGTH-1);
			target.strval[VAR_STRING_LENGTH-1] = '\0';
			break;
		default:
			vnLog(EL_warning, COM_CORE, "setvar :: Unsupported operator '%c' for target type string", op);
			return;
		}

		target.intval = atoi(target.strval);
		break;
	case VT_null:
		//Do nothing
		break;
	}

	vnLog(EL_verbose, COM_SCRIPT, "(g)setvar %s %c %s", name, op, target.strval);
	map[name] = target;
}
